sys/senml/senml.c

```diff
@@ -0,0 +1,429 @@
+                    found_base_info = true;
+                }
+            }
+
+            if (!strcmp(tmp_key, SJ_NAME)) {
+                (pack->records)[array_num].name = tmp_val;
+            }
+            else if (!strcmp(tmp_key, SJ_UNIT)) {
+                (pack->records)[array_num].unit = tmp_val;
+            }
+            else if (!strcmp(tmp_key, SJ_VALUE)) {
+                (pack->records)[array_num].value_type = SENML_TYPE_FLOAT;
+                (pack->records)[array_num].value.f = atof(tmp_val);
+            }
+            else if (!strcmp(tmp_key, SJ_STRING_VALUE)) {
+                (pack->records)[array_num].value_type = SENML_TYPE_STRING;
+                (pack->records)[array_num].value.s = tmp_val;
+            }
+            else if (!strcmp(tmp_key, SJ_BOOL_VALUE)) {
+                (pack->records)[array_num].value_type = SENML_TYPE_BOOL;
+                if (!strcmp(tmp_val, "true")) {
+                    (pack->records)[array_num].value.b = true;
+                }
+                else if (!strcmp(tmp_val, "false")) {
+                    (pack->records)[array_num].value.b = false;
+                }
+            }
+            else if (!strcmp(tmp_key, SJ_DATA_VALUE)) {
+                (pack->records)[array_num].value_type = SENML_TYPE_BINARY;
+                (pack->records)[array_num].value.d = tmp_val;
+            }
+            else if (!strcmp(tmp_key, SJ_VALUE_SUM)) {
+                (pack->records)[array_num].value_sum = atof(tmp_val);
+            }
+            else if (!strcmp(tmp_key, SJ_TIME)) {
+                (pack->records)[array_num].time = atof(tmp_val);
+            }
+            else if (!strcmp(tmp_key, SJ_UPDATE_TIME)) {
+                (pack->records)[array_num].update_time = atof(tmp_val);
+            }
+        }
+    }
+
+    if (!found_base_info) {
+        pack->base_info = NULL;
+    }
+
+    pack->num = (array_num + 1);
+
+    return 0;
+}
+
+
+int senml_encode_json_s(const senml_pack_t *pack, char *output, size_t len)
+{
+    if (len < 3) {           // we need space for at least '[', ']', and '\0'
+        return -1;
+    }
+
+    size_t ins_pos = 0;      // position in output where next character will be inserted
+    size_t buf_len = len;    // length of space left in output
+
+    size_t ins_pos_after_bi; // stores value of ins_pos after inserting base info to check
+                             // whether all records provided were empty
+
+    int chars_written = 0;   // stores return values of snprintf calls: negative value indicates
+                             // error, value >= buf_len indicates insufficient space in output
+
+    bool insert_opening_brace = true;  // is used to check whether '{' has to be inserted after
+                                       // we the first record is empty but base info is provided
+
+    output[ins_pos] = '[';
+    ins_pos++;
+    buf_len--;
+
+    if (!pack->base_info) {
+        goto encode_records;
```

Comment:
I mean putting everything between this and the `encode_records` label into an `if (pack->base_info)` branch would make the code far more readable.